### PR TITLE
#1225 form components

### PR DESCRIPTION
**Pull request by BALEHOK**

### CODE REVIEW

File: .storybook/main.js [hunk 1]
```diff
@@ -23,11 +23,7 @@ const rootDir = path.resolve(__dirname, "../");
 
 module.exports = {
   stories: ["../src/**/*.stories.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx)"],
-  addons: [
-    "@storybook/addon-links",
-    "@storybook/addon-essentials",
-    "@storybook/addon-actions",
```
Comment (BALEHOK) [inline .storybook/main.js:29]: `addon-actions` included in `addon-essentials`

File: src/components/fields/fieldUtils.ts [hunk 2]
```diff
@@ -19,9 +19,3 @@ export function fieldLabel(name: string): string {
   const parts = name.split(".");
   return parts[parts.length - 1];
 }
-
-export function makeInitialValues<InitialType = Record<string, unknown>>(
-  values: InitialType
-): InitialType {
-  return values;
-}
```
Comment (BALEHOK) [inline src/components/fields/fieldUtils.ts:27]: not used

File: src/components/form/Form.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  age: "",
+};
+
+export const WithHorizontalFormGroup: ComponentStory<typeof Form> = (args) => (
+  <Form validationSchema={SchemaShape} initialValues={initialValues} {...args}>
+    <HorizontalFormGroup label="Title" propsOrFieldName="title">
+      {(field, meta) => (
+        <BootstrapForm.Control {...field} isInvalid={Boolean(meta.error)} />
+      )}
+    </HorizontalFormGroup>
```
Comment (BALEHOK) [inline src/components/form/Form.stories.tsx:59]: Current `HorizontalFormGroup`

File: src/components/form/Form.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  </Form>
+);
+WithHorizontalFormGroup.storyName = "With HorizontalFormGroup";
+
+export const WithHorizontalTextField: ComponentStory<typeof Form> = (args) => (
+  <Form validationSchema={SchemaShape} initialValues={initialValues} {...args}>
+    <HorizontalTextField placeholder="Title" name="title" />
```
Comment (BALEHOK) [inline src/components/form/Form.stories.tsx:84]: Simplified version of `HorizontalFormGroup`

File: src/options/pages/installed/ShareExtensionModal.tsx [hunk 5]
```diff
@@ -103,7 +104,7 @@ const ShareExtensionModal: React.FC<{
     public: true,
   };
 
-  const handleShare = useCallback(
+  const handleShare: OnSubmit = useCallback(
```
Comment (BALEHOK) [inline src/options/pages/installed/ShareExtensionModal.tsx:107]: ToDo: discuss usage of `useCallback`.

File: src/components/form/fields/HorizontalTextField.tsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  name: string;
+  label?: string;
+  placeholder?: string;
+  description?: ReactNode;
+  error?: string;
+  touched?: boolean;
+  onBlur?: FocusEventHandler<HTMLInputElement>;
+};
+
+export const HorizontalTextField: React.FC<HorizontalTextFieldProps> = ({
```
Comment (BLoe) [inline src/components/form/fields/HorizontalTextField.tsx:38]: Why do we need to export this?

**Review by BLoe — COMMENTED** [review]
This all looks good to me, other than a few questions I noted here. Let's sync up soon to talk about doing the rest of the control types and cutting over the rest of the forms in the code base.

File: src/components/form/fields/HorizontalTextField.tsx (hunk 6 above)
Comment (BALEHOK) [reply src/components/form/fields/HorizontalTextField.tsx:38]: To use the component outside of Formik context, for instance with JSON Schema Form. Would be nice to see the same UI regardless of the technical implementation. Formik, JSON Schema Form, or any other engine, UI always looks the same.

File: src/components/form/fields/HorizontalTextField.tsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  error,
+  touched,
+  ...restFieldProps
+}) => {
+  const hasError = Boolean(error);
+  return (
+    <BootstrapForm.Group as={Row} controlId={name}>
+      {label && (
+        <BootstrapForm.Label column sm="2">
+          {fieldLabel(label)}
```
[the remark this answers is not in the document]
Comment (BALEHOK) [reply src/components/form/fields/HorizontalTextField.tsx:52]: Good catch! `fieldLabel` should be used outside of the field, it's up to the concrete form implementation to decide what to show.

File: src/components/form/Form.tsx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+}) => (
+  <Formik
+    initialValues={initialValues}
+    validationSchema={validationSchema}
+    validateOnMount={validateOnMount}
+    onSubmit={onSubmit}
+  >
+    {({ handleSubmit, isSubmitting, isValid, values, status }) => (
+      <BootstrapForm noValidate onSubmit={handleSubmit}>
+        {status && <div className="text-danger mb-3">{status}</div>}
```
[the remark this answers is not in the document]
Comment (BALEHOK) [reply src/components/form/Form.tsx:64]: You're right (although often it's very tempting just to apply a BS class).
But the colors should never be hardcoded, need to define the vars for them...

[the remark this answers is not in the document]
Comment (BLoe) [reply src/components/form/Form.tsx:64]: >colors should never be hardcoded

I think it's not a huge deal right now, but agreed that we should define the colors in a central place at some point. Do you want to start a colors file in the `/themes` folder or somewhere else that makes sense?

File: src/components/form/fields/HorizontalTextField.tsx (hunk 6 above)
Comment (BLoe) [reply src/components/form/fields/HorizontalTextField.tsx:38]: Ok, makes sense. Is there going to be confusion here with the names of this versus the default export? I'm seeing a warning already at the import site where you are using the default export. Do you think we should name the non-formik ones something slightly different?

Comment (BALEHOK) [reply src/components/form/fields/HorizontalTextField.tsx]: I pulled Formik component to another file.

File: src/components/form/Form.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (BALEHOK) [reply src/components/form/Form.tsx]: TBD. There's a `vendors/theme/assets/styles` folder and `themes` folder at the root level. I'd prefer to use the latter. `vendors` now is full of imported styles, variables, and mixins, ideally, we need to clean it up.
cc @twschiller

[the remark this answers is not in the document]
Comment (BLoe) [reply src/components/form/Form.tsx]: I think it makes sense to try not to mess with the `/vendors` stuff if at all possible.

**Review by twschiller — APPROVED** [review]
There's still areas of improvements, but let's move forward and figure out how to improve as we go

Areas of improvement:
- [ ] There's basically no type-safety right now. The type is hard to write because there's alternatives. (Link for inspiration for later: https://medium.com/@jrwebdev/react-higher-order-component-patterns-in-typescript-42278f7590fb)
- [ ] Code re-use for vertical forms
- [ ] We now have yet another fields folder in our file structure, we need to reconcile with out other one at `components/fields`

**Pull request merged**